client/src/AppBundle/Service/DocumentSyncService.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace AppBundle\Service;
+
+use AppBundle\Entity\Ndr\Ndr;
+use AppBundle\Entity\Report\Document;
+use AppBundle\Entity\Report\Report;
+use AppBundle\Entity\ReportInterface;
+use AppBundle\Service\File\Storage\S3Storage;
+use DateTime;
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\RequestException;
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\Psr7\Response as Psr7Response;
+use Symfony\Component\HttpFoundation\Response;
+
+class DocumentSyncService
+{
+    const PERMANENT_ERRORS = [
+        Response::HTTP_REQUEST_ENTITY_TOO_LARGE,
+        Response::HTTP_UNSUPPORTED_MEDIA_TYPE,
+    ];
+
+    /** @var S3Storage */
+    private $storage;
+
+    public function __construct(S3Storage $storage)
+    {
+        $this->storage = $storage;
+        $this->client = new Client([
+            'base_uri' => 'http://pact-mock'
+        ]);
+    }
+
+    public function syncReportDocument(Document $document)
+    {
+        $contents = $this->storage->retrieve($document->getStorageReference());
+
+        $report = $document->getReport();
+        $attributes = $this->getAttributesFromReport($report);
+
+        try {
+            $response = $this->sendReportDocument($report->getClient()->getCaseNumber(), $contents, $attributes);
+
+            $data = json_decode(strval($response->getBody()), true);
+            return $data['data']['id'];
+        } catch (RequestException $exception) {
+            $response = $exception->getResponse();
+            if ($response) {
+                $response = json_decode($exception->getResponse()->getBody());
+
+                return $response->errors[0]->detail;
+            } else {
+                return $exception->getMessage();
+            }
+        }
+    }
+
+    /**
+     * @param Report|Ndr $report
+     * @return array
+     */
+    private function getAttributesFromReport(ReportInterface $report): array
+    {
+        if ($report instanceof Ndr) {
+            $type = 'NDR';
+        } else if (in_array($report->getType(), [Report::TYPE_HEALTH_WELFARE, Report::TYPE_COMBINED_HIGH_ASSETS, Report::TYPE_COMBINED_LOW_ASSETS])) {
+            $type = 'HW';
+        } else {
+            $type = 'PF';
+        }
+
+        if ($report instanceof Ndr) {
+            $endDate = $report->getStartDate()->format('Y-m-d');
```

Comment:
Might be my ignorance with NDRs but how does the start date work out to be the end date 🤔

Reply:
NDRs are issued outside of annual reporting, so don't have a reporting period. In practice, they're assigned a start date to match the first report's start date. For the sake of providing a required value, I'm setting the end date to the same date.

Court date: 12th May 2017
NDR: 12th May 2017–12th May 2017
Report 1: 12th May 2017–11th May 2018
Report 2: 12th May 2018–11th May 2019

I'll draw this to Stacey's attention though. Maybe we don't want to require the end date in the API (or not sync NDRs).